### DISCUSSION TITLE
VIM support with official copilot.vim

### DISCUSSION THREAD

**Discussion by petronny** [moyix/fauxpilot#72] [Ideas]
After 3 days of digging, I finally get fauxpilot working with the official vim copilot plugin.

![微信截图_20221002154845](https://user-images.githubusercontent.com/5357170/193443765-50ab4350-37a9-407e-95da-148859a632b0.png)
![微信图片_20221002170119](https://user-images.githubusercontent.com/5357170/193446490-c9a7abfa-be05-4155-953a-699c69643b7b.png)

<s>It's a little bit long. But I will try to make it shorter.</s>

Good news! Things have been fixed in both fauxpilot and copilot.vim. Now the guide is much shorter.

### Configure vim plugin
* Install [github/copilot.vim](https://github.com/github/copilot.vim).
* Run `:Copilot setup` for a basic authentication.
* Go to the directory of copilot.vim and run
```sh
sed 's|https://api.github.com/copilot_internal|http://127.0.0.1:5000/copilot_internal|g' -i copilot/dist/agent.js
sed 's|https://copilot-proxy.githubusercontent.com|http://127.0.0.1:5000|g' -i copilot/dist/agent.js
```
* Restart vim.

### It's done!

Currently I've tested `getCompletions` and `getPanelCompletions`. Please let me know if something not works.

Do not use the proxy setting, `g:copilot_proxy`, in copilot.vim.  
It works like a conventional http proxy, not just replacing `https://copilot-proxy.githubusercontent.com`.
I don't think it's easy to hijack https requests so don't use it.

The agent is also sending information to copilot telemetry.  
This may get you banned if you have no copilot subscription.
Need to change the telemery server to the local server in the future.

### Capture the inputs and outputs of the agent

Capturing the inputs and outputs of the agent is very helpful for debugging.

* Create a script called `agent.sh`.
```sh
tee /tmp/agent.in | node /path/to/copilot.vim/copilot/dist/agent.js | tee /tmp/agent.out
```
* Change `g:copilot_node_command`, `g:copilot_agent_command` and `g:copilot_ignore_node_version`.
```vim
let g:copilot_node_command = 'sh'
let g:copilot_ignore_node_version = 1
let g:copilot_agent_command= '/path/to/your/agent.sh'
```
* Then see `/tmp/agent.in` and `/tmp/agent.out`.

### Implement a custom agent

Although the shipped agent is in javascript, you actually can implement it with any language you like.
It's a jsonrpc server based on STDIO. Just update `agent.sh` to run your custom agent.
```sh
tee /tmp/agent.in | python custom-agent.py | tee /tmp/agent.out
```
There are other patches you may need to develope a custom agent.
```patch
diff --git a/autoload/copilot.vim b/autoload/copilot.vim
index 58122a4..c634312 100644
--- a/autoload/copilot.vim
+++ b/autoload/copilot.vim
@@ -543,7 +543,7 @@ function! s:VerifySetup() abort
     return
   endif

-  let status = copilot#Call('checkStatus', {})
+  let status = copilot#Call('checkStatus', {'options': {'localChecksOnly': v:true}})

   if !has_key(status, 'user')
     echo 'Copilot: Not authenticated. Invoke :Copilot setup'
@@ -601,7 +601,7 @@ function! s:commands.setup(opts) abort

   let browser = copilot#Browser()

-  let status = copilot#Call('checkStatus', {})
+  let status = copilot#Call('checkStatus', {'options': {'localChecksOnly': v:true}})
   if has_key(status, 'user')
     let data = {}
   else
```

**Comment by moyix** [comment]
Thanks, this is great info! Hopefully we can get your changes merged in quickly so that the experience of using FauxPilot with vim is a lot smoother :)

**Comment by 152334H** [comment]
Hi, is this still being worked on? I notice there's a PR for it at #76 that's been stale for a while.

**Reply by petronny** [reply]
Hi, that PR is not need anymore :)

**Comment by sisrfeng** [comment]
Will this be merged into the main branch?

**Comment by mward25** [comment]
I'm having trouble getting this to work.

 If this is not where I am supposed to ask for help, or if I did something else really stupid please let me know (I haven't communicated a lot online, and I do not know the exact rules of GitHub etiquette).

Anyway, I managed to get it to the point where the Vim Copilot  Plugin does not show any errors, but it wont show any suggestions either. It will still open up the panel though and claim that 10 suggestions have been made, but no results seem to show up (even after typing ]], and [[).

![screenshot85](https://user-images.githubusercontent.com/61633012/214174796-10389f60-1c95-4d1b-b8e1-82a42100c053.png)


It seems that the web client (https://github.com/fauxpilot-addon/fauxpilot-webclient) works perfectly fine though.

![screenshot86](https://user-images.githubusercontent.com/61633012/214183800-f54a6106-5988-487d-9504-3e4c7821c651.png)


Here is the diff of fauxpilot:
```diff
diff --git a/copilot_
proxy/app.py b/copilot_proxy/app.py
index 81f25f2..87a6623 100644
--- a/copilot_proxy/app.py
+++ b/copilot_proxy/app.py
@@ -5,6 +5,7 @@ import uvicorn
 from fastapi import FastAPI, Request, Response
 from fastapi.responses import JSONResponse
 from sse_starlette.sse import EventSourceResponse
+from fastapi.responses import StreamingResponse
 
 from config.log_config import uvicorn_logger
 from models import OpenAIinput
@@ -49,9 +50,9 @@ async def completions(data: OpenAIinput):
         )
 
     if data.get("stream") is not None:
-        return EventSourceResponse(
-            content=content,
+        return StreamingResponse(
             status_code=200,
+            content=codegen(data=data),
             media_type="text/event-stream"
         )
     else:
diff --git a/copilot_proxy/models.py b/copilot_proxy/models.py
index 8dd6b94..83a6302 100644
--- a/copilot_proxy/models.py
+++ b/copilot_proxy/models.py
@@ -2,9 +2,10 @@ from typing import Optional, Union
 
 from pydantic import BaseModel, constr
 
+# constr(regex="^(fastertransformer|py-model)$") = "fastertransformer"
 
 class OpenAIinput(BaseModel):






-    model: constr(regex="^(fastertransformer|py-model)$") = "fastertransformer"
+    model: Optional[str] = "fastertransformer"
     prompt: Optional[str]
     suffix: Optional[str]
     max_tokens: Optional[int] = 16
```

And here is the diff of the copilot vim plugin (the do_sed_stuff file is the one I execute to modify agent.js):
```diff
diff --git a/autoload/copilot.vim b/autoload/copilot.vim
index c8fd8eb..0b29fea 100644
--- a/autoload/copilot.vim
+++ b/autoload/copilot.vim
@@ -545,7 +545,7 @@ function! s:VerifySetup() abort
     return
   endif
 
-  let status = copilot#Call('checkStatus', {})
+  let status = copilot#Call('checkStatus', {'options': {'localChecksOnly': v:true}})
 
   if !has_key(status, 'user')
     echo 'Copilot: Not authenticated. Invoke :Copilot setup'
@@ -609,7 +609,7 @@ function! s:commands.setup(opts) abort
 
   let browser = copilot#Browser()
 
-  let status = copilot#Call('checkStatus', {})
+  let status = copilot#Call('checkStatus', {'options': {'localChecksOnly': v:true}})
   if has_key(status, 'user')
     let data = {}
   else
diff --git a/copilot/dist/agent.js b/copilot/dist/agent.js
index 58c599b..d89ca24 100644
Binary files a/copilot/dist/agent.js and b/copilot/dist/agent.js differ
diff --git a/do_sed_stuff b/do_sed_stuff
new file mode 100755
index 0000000..a978f04
--- /dev/null
+++ b/do_sed_stuff
@@ -0,0 +1,15 @@
+sed 's|https://api.github.com/copilot_internal|http://127.0.0.1:5001/copilot_internal|g' -i copilot/dist/agent.js
+sed 's|https://copilot-proxy.githubusercontent.com|http://localhost:5000|g' -i copilot/dist/agent.js
+sed 's|/v1/engines/copilot-codex|/v1/engines/codegen|' -i copilot/dist/agent.js
+
+#sed 's|/v1/engines/copilot-codex|/v1/engines/fastertransformer|' -i copilot/dist/agent.js
+
+# print five ints
+
+#sed 's|https://api.github.com/copilot_internal|http://localhost:5001/copilot_internal|g' -i copilot/dist/agent.js
+#sed 's|https://copilot-proxy.githubusercontent.com|http://localhost:5000|g' -i copilot/dist/agent.js
+#sed 's|/v1/engines/copilot-codex|/v1/engines/codegen|' -i copilot/dist/agent.js
+
+#sed 's|https://api.github.com/copilot_internal|http://localhost:8001/copilot_internal|g' -i copilot/dist/agent.js
+#sed 's|https://copilot-proxy.githubusercontent.com|http://localhost:8000|g' -i copilot/dist/agent.js
+#sed 's|/v1/engines/copilot-codex|/v1/engines/codegen|' -i copilot/dist/agent.js
```

Thank you.

**Reply by mward25** [reply]
I got it working, instead of using the github copilot plugin, I used https://github.com/tom-doerr/vim_codex and modified it to work with fauxpilot. 

Here is the diff for the only file I modified (~/.vim/bundle/vim_codex/python/plugin.py):
```diff
import openai
 
 MAX_SUPPORTED_INPUT_LENGTH = 4096
-USE_STREAM_FEATURE = True
+USE_STREAM_FEATURE = False
 MAX_TOKENS_DEFAULT = 64
 
 CONFIG_DIR = os.getenv('XDG_CONFIG_HOME', os.path.expanduser('~/.config'))
@@ -29,7 +29,7 @@ def create_template_ini_file():
         with open(API_KEYS_LOCATION, 'w') as f:
             f.write('[openai]\n')
             f.write('organization_id=\n')
-            f.write('secret_key=\n')
+            f.write('secret_key=dummy\n')
 
         print('OpenAI API config file created at {}'.format(API_KEYS_LOCATION))
         print('Please edit it and add your organization ID and secret key')
@@ -44,17 +44,18 @@ def initialize_openai_api():
     Initialize the OpenAI API
     """
     # Check if file at API_KEYS_LOCATION exists
-    create_template_ini_file()
-    config = configparser.ConfigParser()
-    config.read(API_KEYS_LOCATION)
+    #create_template_ini_file()
+    #config = configparser.ConfigParser()
+    #config.read(API_KEYS_LOCATION)
 
-    openai.organization_id = config['openai']['organization_id'].strip('"').strip("'")
-    openai.api_key = config['openai']['secret_key'].strip('"').strip("'")
+    #openai.organization_id = config['openai']['organization_id'].strip('"').strip("'")
+    #openai.api_key = config['openai']['secret_key'].strip('"').strip("'")
 
 
-def complete_input_max_length(input_prompt, max_input_length=MAX_SUPPORTED_INPUT_LENGTH, stop=None, max_tokens=64):
+def complete_input_max_length(input_prompt, max_input_length=MAX_SUPPORTED_INPUT_LENGTH, stop="\n\n", max_tokens=64):
     input_prompt = input_prompt[-max_input_length:]
-    response = openai.Completion.create(engine='code-davinci-001', prompt=input_prompt, best_of=1, temperature=0.5, max_tokens=max_tokens, stream=USE_STREAM_FEATURE, stop=stop)
+    #response = openai.Completion.create(engine='codegen', prompt=input_prompt, best_of=1, temperature=0.5, max_tokens=max_tokens, stream=USE_STREAM_FEATURE, stop=stop)
+    response = openai.Completion.create(engine='codegen', prompt=input_prompt, temperature=0.5, max_tokens=max_tokens, stop=stop)
     return response
 
 def complete_input(input_prompt, stop, max_tokens):
@@ -121,10 +122,9 @@ def get_first_line_below_cursor_with_text():
-     def create_completion(stop=None): 
+     def create_completion(stop="\n\n"): 
     try:
-        from AUTH import ORGANIZATION_ID, SECRET_KEY
-
-        openai.organization = ORGANIZATION_ID
-        openai.api_key = SECRET_KEY
+        #openai.organization = "v1"
+        openai.api_key = 'dummy'
+        openai.api_base = 'http://0.0.0.0:5000/v1'
     except ModuleNotFoundError:
         initialize_openai_api()
     max_tokens = get_max_tokens()
```

**Reply by sisrfeng** [reply]
Could you share your entire repo on this topic?

**Reply by petronny** [reply]
I take a look today and find that some patches are not needed any more. Please try the updated guide.

**Reply by mward25** [reply]
I tried the guide (Btw, I re-cloned fauxpilot and copilot.vim from scratch, so both are running on fresh setups (minus the sed commands I executed from the guide of course)). But I received this error. Last time I tried to follow the guide I received a similar, if not identical error, and I fixed it last time, but I made a bunch of changes and I don't know what caused the error to go away:
```
Error detected while processing function copilot#Command[38]..function copilot#Command[30]..12[1]..<SNR>29_VerifySetup[7]..copilot#Call[2]..<SNR>117_AgentCall[5]..<SNR>117_RequestAwait:
line    5:
E605: Exception not caught: copilot#agent(-32603): Request checkStatus failed with message: connect ECONNREFUSED ::1:5001
```
EDIT: I realized that something I wrote was not quite true, so I corrected it.
EDIT: Fixed formatting

**Reply by mward25** [reply]
> Could you share your entire repo on this topic?

I created a repo with my changes in it, is this what you mean?
https://github.com/mward25/vim_codex_fauxpilot

EDIT: Just wanted to make a quick note that this repo works with unedited Fauxpilot source code.

**Reply by petronny** [reply]
> Error detected while processing function copilot#Command[38]..function copilot#Command[30]..12[1]..29_VerifySetup[7]..copilot#Call[2]..117_AgentCall[5]..117_RequestAwait:
line 5:
E605: Exception not caught: copilot#agent(-32603): Request checkStatus failed with message: connect ECONNREFUSED ::1:5001

Seems that the fake copilot authentication server is not started at localhost:5001.

**Reply by mward25** [reply]
That's weird, I started the flask server(with flask run --port 5001), here is the output from the flask server:

```
 * Debug mode: off
WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://127.0.0.1:5001
Press CTRL+C to quit
```

**Reply by mward25** [reply]
I figured it out, the ip addresses are not the same on all computers. Running the following sed commands made it work for me:

```
sed 's|https://api.github.com/copilot_internal|http://127.0.0.1:5001/copilot_internal|g' -i copilot/dist/agent.js
sed 's|https://copilot-proxy.githubusercontent.com|http://0.0.0.0:5000|g' -i copilot/dist/agent.js
sed 's|/v1/engines/copilot-codex|/v1/engines/codegen|' -i copilot/dist/agent.js
```

EDIT: Sorry if this is too obvious to other people, but I thought I should mention that I found the 0.0.0.0 adress from fauxpilots output on the line that said:
```
Attaching to fauxpilot-copilot_proxy-1, fauxpilot-triton-1
fauxpilot-copilot_proxy-1  | INFO:     Started server process [1]
fauxpilot-copilot_proxy-1  | INFO:     Waiting for application startup.
fauxpilot-copilot_proxy-1  | INFO:     Application startup complete.
fauxpilot-copilot_proxy-1  | INFO:     Uvicorn running on http://0.0.0.0:5000 (Press CTRL+C to quit)
```

**Reply by petronny** [reply]
localhost is resolved to ::1, so `flask run --host ::1 --port 5001`.
Maybe forcing everything to `127.0.0.1` is better. Guide updated.

**Reply by mward25** [reply]
Thanks.

**Comment by petronny** [comment]
Guide updated.

**Reply by sisrfeng** [reply]
大佬, 有没有知乎或者b站账号啊, 校友渣渣想向大佬学习~

**Reply by petronny** [reply]
I do have account on zhihu and bilibili with the same username but I rarely post things.

And please don't create off-topic replies.

**Comment by nick-ohi** [comment]
Great work on this! Exactly what I am looking for as a vim user.

In the guide you say:
> The agent is also sending information to copilot telemery.
> This may get you banned if you have no copilot subscription.
> Need to change the telemery server to the local server in the future.

Any ideas on what it takes to change this?

Thanks!

**Reply by petronny** [reply]
Here is the way but I still don't have time to do it:
1. Set `g:copilot_proxy` to a http proxy (eg. `mitmproxy`) to show all the internet connections.
2. Find those urls which are something like `https://copilot-telemetry.github.com/whatever`. I don't remember the urls exactly.
3. Replace the urls with `http://127.0.0.1:5000/whatever` in `agent.js`.
4. If the plugin is not working, implement the responses for these telemetry connections in `copilot_proxy/app.py`.

**Reply by xjtu-blacksmith** [reply]
>     2. Find those urls which are something like `https://copilot-telemetry.github.com/whatever`. I don't remember the urls exactly.

I've searched in `agent.js` and it reads `https://copilot-telemetry.githubusercontent.com/telemetry`, which could be replaced.

**Comment by GregBowyer** [comment]
Anyone figured out how to replace the tokenizers?

**Comment by BPplays** [comment]
@petronny this guide seems to be outdated is there any way to get this working now?
